src/Functions/sparseGrams.cpp

```diff
@@ -0,0 +1,287 @@
+#include <optional>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/FunctionTokens.h>
+#include <Common/Exception.h>
+
+#include <zlib.h>
+#include <Poco/UTF8Encoding.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+/** Functions that finds all substrings win minimal length n
+  * such their border (n-1)-grams' hashes are more than hashes of every (n-1)-grams' in substring.
+  * As a hash function use zlib crc32, which is crc32-ieee with 0xffffffff as initial value
+  *
+  * sparseGrams(s)
+  */
+namespace
+{
+
+using Pos = const char *;
+
+template <bool is_utf8>
+class SparseGramsImpl
+{
+private:
+    Pos pos;
+    Pos end;
+    std::vector<UInt32> ngram_hashes;
+    std::vector<size_t> utf8_offsets;
+    size_t left;
+    size_t right;
+    UInt64 min_ngram_length = 3;
+    UInt64 max_ngram_length = 100;
+
+    void buildNgramHashes()
+    {
+        if constexpr (is_utf8)
+        {
+            Poco::UTF8Encoding encoder{};
+            size_t byte_offset = 0;
+            while (pos + byte_offset < end)
+            {
+                utf8_offsets.push_back(byte_offset);
+                auto len = encoder.sequenceLength(reinterpret_cast<const unsigned char *>(pos + byte_offset), end - pos - byte_offset);
+                if (len < 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Incorrect utf8 symbol");
+                byte_offset += len;
+            }
+            if (pos + byte_offset != end)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Incorrect utf8 symbol");
+
+            utf8_offsets.push_back(byte_offset);
+
+            if (utf8_offsets.size() >= min_ngram_length)
+                ngram_hashes.reserve(utf8_offsets.size() - min_ngram_length + 1);
+            for (size_t i = 0; i + min_ngram_length - 1 < utf8_offsets.size(); ++i)
+                ngram_hashes.push_back(crc32_z(
+                    0UL,
+                    reinterpret_cast<const unsigned char *>(pos + utf8_offsets[i]),
+                    utf8_offsets[i + min_ngram_length - 1] - utf8_offsets[i]));
+        }
+        else
+        {
+            if (pos + min_ngram_length <= end)
+                ngram_hashes.reserve(end - pos - min_ngram_length + 1);
+            for (size_t i = 0; pos + i + min_ngram_length - 2 < end; ++i)
+                ngram_hashes.push_back(crc32_z(0L, reinterpret_cast<const unsigned char *>(pos + i), min_ngram_length - 1));
+        }
+    }
+
+    std::optional<std::pair<size_t, size_t>> getNextIndices()
+    {
+        chassert(right > left);
+        while (left < ngram_hashes.size())
+        {
+            while (right < ngram_hashes.size() && right <= left + max_ngram_length - min_ngram_length + 1)
```

Comment:
The size is calculated in a loop, and the compiler typically cannot optimize it.